### DISCUSSION TITLE
Private IP Address VS Public IP Address

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#210] [Network]
.

**Comment by nayonsoso** [comment]
### Public IP (공인 IP)
- 전체 인터넷 상에서 유일하다.
- 외부에 공개되어있다.
    - 따라서 다른 PC로부터 접근이 가능하기 때문에 방화벽 등의 보안 프로그램을 설치할 필요가 있다.
- ISP(인터넷 서비스 공급자 e.g. KT, SKT)가 할당한다.
 
cf. 더 구체적으로는, 전 세계적으로 ICANN이라는 기관이 국가별로 사용할 IP 대역을 관리하고, 우리나라는 한국인터넷진흥원(KISA)에서 국내 IP 주소들을 관리하는데, 이것을 ISP(Internet Service Provider의 약자로 KT, LG, SKT와 같이 인터넷을 제공하는 통신업체)가 부여받고, 우리는 위 회사에 가입을 통해 IP를 제공받아 인터넷을 사용하게 되는 것이다.

### Private IP (사설 IP, 로컬 IP)
- private network 에서만 사용할 수 있다.
- private network 에서만 유일하다.
- 외부에서 접속할 수 없다.
- 라우터가 개인 장치에 할당한다.
- 범위가 정해져있다.
    - 10.0.0.0 ~ 10.255.255.255
    - 172.16.0.0 ~ 172.31.255.255
    - 192.168.0.0. ~ 192.168.255.255
- 사설 IP 덕분에 한정된 공용 IPv4 주소를 아낄 수 있으며, 네트워크 내부의 통신을 더 안전하게 할 수 있다.

| **구분**               | **공인 IP (Public IP)**                 | **사설 IP (Private IP)**            |
|------------------------|-----------------------------------------|-------------------------------------|
| **할당 주체**          | ISP(인터넷 서비스 공급자)               | 라우터(공유기)                      |
| **할당 대상**          | 개인 또는 회사의 서버(라우터)           | 개인 또는 회사의 기기               |
| **고유성**             | 인터넷 상에서 유일한 주소               | 하나의 네트워크 안에서 유일         |
| **공개 여부**          | 내/외부 접근 가능                       | 외부 접근 불가능                    |

---


### 사설 IP를 사용하는 우리는 어떻게 인터넷에 접속할 수 있나?🧐
앞서 말했든, 사설 IP는 외부에서 알 수 없다. 그럼 우리가 인터넷에 보내는 요청을 응답해줄 때, 인터넷의 요청은 어떻게 우리 컴퓨터까지 닿을 수 있는것일까? 우리 컴퓨터의 사설 IP를 모를텐데!


### NAT
이때 사용되는 기술이 NAT(Network Address Translation)이다. 우리는 공유기로부터 할당받은 사설 IP를 가지고 있고, 공유기는 ISP에 할당받은 공용 IP를 가지고 있다. 내 컴퓨터에서 어떤 공용 IP에 요청을 보낼 때, 공유기를 거쳐가는데, 이때 내 컴퓨터의 사설 IP를 공유기의 공용 IP로 변환해서 요청을 보낸다. 그리고 외부에서 들어오는 응답을 공유기가 내 컴퓨터의 사설 IP로 보내서 통신을 완료한다.

ref
https://inpa.tistory.com/entry/WEB-🌐-IP-기초-사설IP-공인IP-NAT-개념-정말-쉽게-정리
https://inpa.tistory.com/entry/WEB-🌐-NAT-란-무엇인가
https://velog.io/@hidaehyunlee/공인Public-사설Private-IP의-차이점

**Comment by gyungchan-jo** [comment]
### 퍼블릭 IP 주소 (Public IP Address)

퍼블릭 IP 주소는 인터넷에 직접 연결된 장치에 할당되는 IP 주소로, 전 세계 어디서나 접근할 수 있습니다.

### 프라이빗 IP 주소 (Private IP Address)

프라이빗 IP 주소는 내부 네트워크에서 사용되는 IP 주소로, 외부 인터넷에서는 접근할 수 없습니다.

---

차이 : 
1. 주소 범위
퍼블릭 IP 주소는 전 세계적으로 고유하며, 중복되지 않도록 관리됩니다. ICANN (Internet Corporation for Assigned Names and Numbers)에서 관리합니다. 프라이빗 IP 주소는 아래와 같이 범위가 정해져있습니다.
    - 10.0.0.0 ~ 10.255.255.255
    - 172.16.0.0 ~ 172.31.255.255
    - 192.168.0.0 ~ 192.168.255.255

2. 용도
퍼블릭 IP 주소는 웹 서버, 이메일 서버 등 외부 인터넷에서 접근해야 하는 장치에 할당됩니다. 인터넷 서비스 제공자(ISP)로부터 할당받습니다.
프라이빗 IP 주소는 내부 네트워크 내에서 장치 간 통신을 위해 사용됩니다. 예를 들어, 가정 내의 컴퓨터, 스마트폰, 프린터 등이 프라이빗 IP 주소를 사용합니다.

3. 보안
퍼블릭 IP 주소는 외부 인터넷에 노출되기 때문에 보안 위협에 취약할 수 있으며, 방화벽 등을 통해 보호가 필요합니다.
프라이빗 IP 주소는 외부 인터넷에서 직접 접근할 수 없기 때문에 상대적으로 보안이 높습니다.

4. 공개 여부
프라이빗 IP 주소는 내부 네트워크에서만 사용되며, 외부 인터넷에서는 접근할 수 없습니다. NAT를 통해 퍼블릭 IP 주소로 변환되어 인터넷과 통신합니다.
